dashboard/app/controllers/api/v1/assessments_controller.rb

```diff
@@ -79,139 +79,190 @@ def index
+      timestamp: timestamp,
+      level_results: level_results
+    }
+
+    [level_group.id, assessment_response]
+  end
 
-          if level_response
-            case level
-            when TextMatch, FreeResponse
-              student_result = level_response["result"]
-              level_result[:student_result] = student_result
-              level_result[:status] = ""
-            when Multi
-              answer_indexes = Multi.find_by_id(level.id).correct_answer_indexes_array
-              student_result = level_response["result"].split(",").map(&:to_i).sort
-              level_result[:student_result] = student_result
-
-              if student_result == [-1]
-                level_result[:student_result] = []
-                level_result[:status] = "unsubmitted"
-              # Deep comparison of arrays of indexes
-              elsif student_result.present? && student_result - answer_indexes == [] && answer_indexes.length == student_result.length
-                multi_count_correct += 1
-                level_result[:status] = "correct"
-              else
-                level_result[:status] = "incorrect"
-              end
-            when Match
-              student_result = level_response["result"].split(",", -1)
-              # If a student did not answer some of the matching question we will record that as nil
-              student_result = student_result.map do |result|
-                result.empty? ? nil : result.to_i
-              end
-              level_result[:student_result] = student_result
-              option_status = []
-              number_correct = 0
-              student_result.each_with_index do |answer, index|
-                if answer.nil?
-                  option_status[index] = "unsubmitted"
-                else
-                  option_status[index] = "submitted"
-                  if answer == index
-                    number_correct += 1
-                  end
-                end
-              end
-              if number_correct == student_result.length
-                match_count_correct += 1
-              end
-              level_result[:status] = option_status
-            end
+  # Summarizes the given level and student answer into the format expected by
+  # section_responses. Also updates the relevant counts in the given stats hash.
+  private def summarize_level_result(level, student_answer, stats)
+    level_result = {}
+
+    case level
+    when TextMatch, FreeResponse
+      level_result[:type] = "FreeResponse"
+    when Multi
+      level_result[:type] = "Multi"
+      stats[:multi_count] += 1
+    when Match
+      level_result[:type] = "Match"
+      stats[:match_count] += 1
+    end
+
+    if student_answer
+      case level
+      when TextMatch, FreeResponse
+        level_result[:student_result] = student_answer
+        level_result[:status] = ""
+      when Multi
+        answer_indexes = Script.cache_find_level(level.id).correct_answer_indexes_array
```

Comment:
Would it make sense to have a method in the `Multi.rb` model where it takes a student answer and returns the result (`unsubmitted`, `correct`, `incorrect`)?

Reply:
That's a good idea, but I think if we did that, we'd want to also move the logic for the other level types into their respective classes and that feels out of scope for this change.  Ok if we save that for another day?